graphix/channels.py

```diff
@@ -0,0 +1,208 @@
+import numpy as np
+
+from graphix.linalg_validations import check_data_dims, check_data_normalization, check_data_values_type, check_rank
+from graphix.ops import Ops
+
+
+class KrausChannel:
+    """quantum channel class in the Kraus representation.
+    Defined by Kraus operators $K_i$ with scalar prefactors ('coef') $c_i$,
+    where the channel act on density matrix as
+    .. math::
+        \rho' = \sum K_i^\dagger \rho K_i
+
+    the data should satisfy
+    .. math::
+        \sum K_i^\dagger K_i = I
+
+    Attributes
+    ----------
+    nqubit : int
+        number of qubits acted on by the Kraus operators
+    size : int
+        number of Kraus operators (== Choi rank)
+    kraus_ops : array_like(dict())
+        the data in format
+        array_like(dict): [{coef: scalar, operator: array_like}, {coef: scalar, operator: array_like}, ...]
+
+    Returns
+    -------
+    Channel object
+        containing the corresponding Kraus operators
+
+    """
+
+    def __init__(self, kraus_data):
+        """
+        Parameters
+        ----------
+        kraus_data : array_like
+            array of Kraus operator data.
+            array_like(dict): [{coef: scalar, operator: array_like}, {parameter: scalar, operator: array_like}, ...]
+            only works for square Kraus operators
+
+        Raises
+        ------
+        ValueError
+            If empty array_like is provided.
+        """
+
+        # check there is data
+        if not kraus_data:
+            raise ValueError("Cannot instantiate the channel with empty data.")
+
+        if not isinstance(kraus_data, (list, np.ndarray, tuple)):
+            raise TypeError(f"The data must be a list, a numpy.ndarray or a tuple not a {type(kraus_data)}.")
+
+        # check that data is correctly formatted before assigning it to the object.
+        assert check_data_values_type(kraus_data)
+        assert check_data_dims(kraus_data)
+
+        # check that the channel is properly normalized i.e
+        # \sum_K_i^\dagger K_i = Identity
+        assert check_data_normalization(kraus_data)
+
+        self.nqubit = int(np.log2(kraus_data[0]["operator"].shape[0]))
+        self.kraus_ops = kraus_data
+
+        # np.asarray(data, dtype=np.complex128)
+        # number of Kraus operators in the Channel
+        assert check_rank(kraus_data)
```

Comment:
This is better since, if `python assert check_rank(kraus_data)` fails, the first two lines are meaningless.

```python
        assert check_rank(kraus_data)
        self.nqubit = int(np.log2(kraus_data[0]["operator"].shape[0]))
        self.kraus_ops = kraus_data
```

Reply:
Ok. We'll also change all dimensions check for something more efficient soon.